sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala

```diff
@@ -119,4 +119,168 @@ object IntervalUtils {
       case _: IllegalArgumentException => null
     }
   }
+
+  private object ParseState extends Enumeration {
+    val PREFIX,
+        BEGIN_VALUE,
+        PARSE_SIGN,
+        POSITIVE_VALUE_TO_LONG,
+        NEGATIVE_VALUE_TO_LONG,
+        BEGIN_UNIT_NAME,
+        UNIT_NAME_SUFFIX,
+        END_UNIT_NAME = Value
+  }
+  private final val intervalStr = UTF8String.fromString("interval")
+  private final val yearStr = UTF8String.fromString("year")
+  private final val monthStr = UTF8String.fromString("month")
+  private final val weekStr = UTF8String.fromString("week")
+  private final val dayStr = UTF8String.fromString("day")
+  private final val hourStr = UTF8String.fromString("hour")
+  private final val minuteStr = UTF8String.fromString("minute")
+  private final val secondStr = UTF8String.fromString("second")
+  private final val millisStr = UTF8String.fromString("millisecond")
+  private final val microsStr = UTF8String.fromString("microsecond")
+
+  def stringToInterval(input: UTF8String): Option[CalendarInterval] = {
```

Comment:
Why did you use `Option` here? If this method has the same return type with `safeFromString`, we could make diff in `Cast.scala` much less.

Reply:
nit: `safeFromString ` -> `safeAndFastFromString`? Also, could you leave some comments here like `This is the fast version of safeFromString...`?

Reply:
> Why did you use Option here?

The same reason as `stringToDate` and `stringToTimestamp` return `Option[...]`. I believe all 3 functions should follow similar convention - `None` for errors and `Some[]` for valid values.

Returning `null` instead of `None` looks slightly ugly in Scala but this should produce less garbage. Let me think of ...

Reply:
> nit: safeFromString -> safeAndFastFromString?

I would remain the same name for consistency with `stringToDate` and `stringToTimestamp`, or rename it to `safeFromUTF8String`.